src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,144 @@
+        }
+    }
+
+    private boolean isFull() {
+        return size == elements.length;
+    }
+
+    private boolean isIndexValid(int index) {
+        return index < size && index >= 0;
+    }
+
+    private boolean isIndexValidForAdd(int index) {
+        return index <= size && index >= 0;
+    }
+
     @Override
     public void add(T value) {
-
+        elements[size] = value;
+        size++;
+        if (isFull()) {
+            ensureCapacity();
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (!isIndexValidForAdd(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You tried to add element to non exist index"
+            );
+        }
+
+        for (int i = size; i > index; i--) {
+            elements[i] = elements[i - 1];
+        }
+        elements[index] = value;
+        size++;
 
+        if (isFull()) {
+            ensureCapacity();
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            elements[size] = list.get(i);
+            size++;
+            if (isFull()) {
+                ensureCapacity();
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (!isIndexValid(index)) {
+            throw new ArrayListIndexOutOfBoundsException("You"
+                    + " tried to get element by non existed index");
+        }
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (!isIndexValid(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You tried to set element to non exist index"
+            );
+        }
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (!isIndexValid(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You tried to remove element from non exist index"
+            );
+        }
+        final T elementToRemove = elements[index];
+        T[] newElements = (T[]) new Object[elements.length];
+        size--;
+        copyArray(elements, 0, newElements, 0, index);
+        copyArray(elements, index + 1, newElements, index, size - index);
+        elements = newElements;
+        return elementToRemove;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        boolean isElementFound = false;
+        int indexOfFoundElement = 0;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(element, elements[i])) {
+                isElementFound = true;
+                indexOfFoundElement = i;
+            }
+        }
+        if (isElementFound) {
+            return remove(indexOfFoundElement);
+        } else {
+            throw new NoSuchElementException();
+        }
```

Comment:
You are iterating through the array until the end even after you have found the element. It would be better to stop the loop after the element is found and remove it.